From 226827593c678d92cf4b53f06637383552a31290 Mon Sep 17 00:00:00 2001
From: Raul E Rangel <ismell@ismell.org>
Date: Thu, 3 Dec 2020 10:56:46 -0700
Subject: [PATCH 2/2] Add support for running as an Illumos LX zone

These changes allow OpenRC to boot as an Illumos/SmartOS LX zone.

This was mostly a copy/paste of LXC.

Signed-off-by: Raul E Rangel <ismell@ismell.org>
---
 etc/rc.conf                 | 1 +
 init.d/binfmt.in            | 2 +-
 init.d/bootmisc.in          | 2 +-
 init.d/consolefont.in       | 2 +-
 init.d/devfs.in             | 2 +-
 init.d/dmesg.in             | 2 +-
 init.d/fsck.in              | 2 +-
 init.d/hostname.in          | 2 +-
 init.d/hwclock.in           | 2 +-
 init.d/keymaps.in           | 2 +-
 init.d/localmount.in        | 2 +-
 init.d/loopback.in          | 2 +-
 init.d/modules.in           | 2 +-
 init.d/mount-ro.in          | 2 +-
 init.d/net-online.in        | 2 +-
 init.d/netmount.in          | 2 +-
 init.d/numlock.in           | 2 +-
 init.d/procfs.in            | 2 +-
 init.d/root.in              | 2 +-
 init.d/save-keymaps.in      | 2 +-
 init.d/save-termencoding.in | 2 +-
 init.d/swap.in              | 2 +-
 init.d/swclock.in           | 2 +-
 init.d/termencoding.in      | 2 +-
 init.d/urandom.in           | 2 +-
 man/openrc-run.8            | 2 ++
 sh/rc-functions.sh          | 2 +-
 src/librc/librc.c           | 4 ++++
 src/librc/rc.h.in           | 1 +
 29 files changed, 33 insertions(+), 25 deletions(-)

diff --git a/etc/rc.conf b/etc/rc.conf
index 123d58e6..624d8c4f 100644
--- a/etc/rc.conf
+++ b/etc/rc.conf
@@ -176,6 +176,7 @@
 # "vserver"        - Linux vserver
 # "xen0"           - Xen0 Domain (Linux and NetBSD)
 # "xenU"           - XenU Domain (Linux and NetBSD)
+# "zone"           - Solaris/Illumos LX zone
 #rc_sys=""
 
 # if  you use openrc-init, which is currently only available on Linux,
diff --git a/init.d/binfmt.in b/init.d/binfmt.in
index 67120eaf..63c9d21f 100644
--- a/init.d/binfmt.in
+++ b/init.d/binfmt.in
@@ -15,7 +15,7 @@ depend()
 {
 	after clock procfs
 	use modules devfs
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/bootmisc.in b/init.d/bootmisc.in
index 43da1869..1aa41978 100644
--- a/init.d/bootmisc.in
+++ b/init.d/bootmisc.in
@@ -125,7 +125,7 @@ migrate_to_run()
 
 clean_run()
 {
-	[ "$RC_SYS" = VSERVER -o "$RC_SYS" = LXC ] && return 0
+	[ "$RC_SYS" = VSERVER -o "$RC_SYS" = LXC -o "$RC_SYS" = ZONE ] && return 0
 	local dir
 	# If / is still read-only due to a problem, this will fail!
 	if ! checkpath -W /; then
diff --git a/init.d/consolefont.in b/init.d/consolefont.in
index d65dd14c..01b76f9b 100644
--- a/init.d/consolefont.in
+++ b/init.d/consolefont.in
@@ -15,7 +15,7 @@ depend()
 {
 	need termencoding
 	after hotplug bootmisc modules
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 }
 
 start()
diff --git a/init.d/devfs.in b/init.d/devfs.in
index 61ba45f8..e2253de6 100644
--- a/init.d/devfs.in
+++ b/init.d/devfs.in
@@ -15,7 +15,7 @@ depend()
 {
 	provide dev-mount
 	before dev
-	keyword -docker -lxc -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -prefix -systemd-nspawn -vserver -zone
 }
 
 mount_dev()
diff --git a/init.d/dmesg.in b/init.d/dmesg.in
index 641f36fe..84a97c76 100644
--- a/init.d/dmesg.in
+++ b/init.d/dmesg.in
@@ -14,7 +14,7 @@ description="Set the dmesg level for a cleaner boot"
 depend()
 {
 	before dev modules
-	keyword -docker -lxc -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/fsck.in b/init.d/fsck.in
index 7052d808..085ace55 100644
--- a/init.d/fsck.in
+++ b/init.d/fsck.in
@@ -17,7 +17,7 @@ depend()
 {
 	after clock
 	use dev clock modules
-	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -timeout -vserver -uml
+	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -timeout -vserver -uml -zone
 }
 
 _abort() {
diff --git a/init.d/hostname.in b/init.d/hostname.in
index 823008fd..d14be92e 100644
--- a/init.d/hostname.in
+++ b/init.d/hostname.in
@@ -14,7 +14,7 @@ description="Sets the hostname of the machine."
 depend()
 {
 	after clock
-	keyword -docker -lxc -prefix -systemd-nspawn
+	keyword -docker -lxc -prefix -systemd-nspawn -zone
 }
 
 start()
diff --git a/init.d/hwclock.in b/init.d/hwclock.in
index 024bff0c..e6ef436c 100644
--- a/init.d/hwclock.in
+++ b/init.d/hwclock.in
@@ -34,7 +34,7 @@ depend()
 	if yesno $clock_adjfile; then
 		use root
 	fi
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 }
 
 setupopts()
diff --git a/init.d/keymaps.in b/init.d/keymaps.in
index 16ddcc96..798e95f8 100644
--- a/init.d/keymaps.in
+++ b/init.d/keymaps.in
@@ -15,7 +15,7 @@ depend()
 {
 	need termencoding
 	after devfs
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 }
 
 start()
diff --git a/init.d/localmount.in b/init.d/localmount.in
index 31684191..9bda0a7c 100644
--- a/init.d/localmount.in
+++ b/init.d/localmount.in
@@ -16,7 +16,7 @@ depend()
 	need fsck
 	use lvm modules root
 	after clock lvm modules root
-	keyword -docker -jail -lxc -prefix -systemd-nspawn -vserver
+	keyword -docker -jail -lxc -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/loopback.in b/init.d/loopback.in
index 70faf2b8..8c924058 100644
--- a/init.d/loopback.in
+++ b/init.d/loopback.in
@@ -14,7 +14,7 @@ description="Configures the loopback interface."
 depend()
 {
 	after clock
-	keyword -jail -prefix -systemd-nspawn -vserver
+	keyword -jail -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/modules.in b/init.d/modules.in
index abaab85b..f901c6d8 100644
--- a/init.d/modules.in
+++ b/init.d/modules.in
@@ -15,7 +15,7 @@ depend()
 {
 	use isapnp
 	provide modules-load
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 find_modfiles()
diff --git a/init.d/mount-ro.in b/init.d/mount-ro.in
index aa7a57cb..b44d16ea 100644
--- a/init.d/mount-ro.in
+++ b/init.d/mount-ro.in
@@ -14,7 +14,7 @@ description="Re-mount filesytems read-only for a clean reboot."
 depend()
 {
 	after killprocs savecache
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/net-online.in b/init.d/net-online.in
index b496e258..23f77844 100644
--- a/init.d/net-online.in
+++ b/init.d/net-online.in
@@ -16,7 +16,7 @@ depend()
 	after modules net
 	need sysfs
 	provide network-online
-	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -uml -vserver
+	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -uml -vserver -zone
 }
 
 get_interfaces()
diff --git a/init.d/netmount.in b/init.d/netmount.in
index 96f5d3e0..b7591585 100644
--- a/init.d/netmount.in
+++ b/init.d/netmount.in
@@ -26,7 +26,7 @@ depend()
 	use afc-client amd openvpn
 	use dns
 	use root
-	keyword -docker -jail -lxc -prefix -systemd-nspawn -vserver
+	keyword -docker -jail -lxc -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/numlock.in b/init.d/numlock.in
index 6c33c95f..5c399dcd 100644
--- a/init.d/numlock.in
+++ b/init.d/numlock.in
@@ -15,7 +15,7 @@ ttyn=${rc_tty_number:-${RC_TTY_NUMBER:-12}}
 
 depend()
 {
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 _setleds()
diff --git a/init.d/procfs.in b/init.d/procfs.in
index edc8ee19..12e12983 100644
--- a/init.d/procfs.in
+++ b/init.d/procfs.in
@@ -16,7 +16,7 @@ depend()
 	after clock
 	use devfs
 	want modules
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/root.in b/init.d/root.in
index 9d053448..362187bc 100644
--- a/init.d/root.in
+++ b/init.d/root.in
@@ -15,7 +15,7 @@ depend()
 {
 	after clock
 	need fsck
-	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/save-keymaps.in b/init.d/save-keymaps.in
index f35097ce..928cfc5e 100644
--- a/init.d/save-keymaps.in
+++ b/init.d/save-keymaps.in
@@ -13,7 +13,7 @@ depend()
 {
 	need termencoding
 	after bootmisc clock keymaps
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 }
 
 start()
diff --git a/init.d/save-termencoding.in b/init.d/save-termencoding.in
index 9e640dcc..76f3531d 100644
--- a/init.d/save-termencoding.in
+++ b/init.d/save-termencoding.in
@@ -14,7 +14,7 @@ ttyn=${rc_tty_number:-${RC_TTY_NUMBER:-12}}
 
 depend()
 {
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 	use root
 	after bootmisc clock termencoding
 }
diff --git a/init.d/swap.in b/init.d/swap.in
index c40b263d..afc5594a 100644
--- a/init.d/swap.in
+++ b/init.d/swap.in
@@ -13,7 +13,7 @@ depend()
 {
 	after clock root
 	before localmount
-	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -vserver
+	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -vserver -zone
 }
 
 start()
diff --git a/init.d/swclock.in b/init.d/swclock.in
index b5556eb5..e660aa2c 100644
--- a/init.d/swclock.in
+++ b/init.d/swclock.in
@@ -14,7 +14,7 @@ description="Sets the local clock to the mtime of a given file."
 depend()
 {
 	provide clock
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 }
 
 # swclock is an OpenRC built in
diff --git a/init.d/termencoding.in b/init.d/termencoding.in
index 41ddc342..0808343a 100644
--- a/init.d/termencoding.in
+++ b/init.d/termencoding.in
@@ -16,7 +16,7 @@ ttyn=${rc_tty_number:-${RC_TTY_NUMBER:-12}}
 
 depend()
 {
-	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu
+	keyword -docker -lxc -openvz -prefix -systemd-nspawn -uml -vserver -xenu -zone
 	after devfs
 }
 
diff --git a/init.d/urandom.in b/init.d/urandom.in
index 91b54f3a..d9b7c931 100644
--- a/init.d/urandom.in
+++ b/init.d/urandom.in
@@ -16,7 +16,7 @@ depend()
 {
 	after clock
 	need localmount
-	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn
+	keyword -docker -jail -lxc -openvz -prefix -systemd-nspawn -zone
 }
 
 save_seed()
diff --git a/man/openrc-run.8 b/man/openrc-run.8
index ec4b88de..e461ed28 100644
--- a/man/openrc-run.8
+++ b/man/openrc-run.8
@@ -320,6 +320,8 @@ Same as -jail, but for Xen DOM0 systems.
 Same as -jail, but for Xen DOMU systems.
 .It Dv -docker
 Same as -jail, but for docker systems.
+.It Dv -zone
+Same as -jail, but for Illumos LX zones.
 .It Dv -containers
 Same as -jail, but for all relevant container types on the operating
 system.
diff --git a/sh/rc-functions.sh b/sh/rc-functions.sh
index d77c5dd3..27c257b7 100644
--- a/sh/rc-functions.sh
+++ b/sh/rc-functions.sh
@@ -98,7 +98,7 @@ _get_containers() {
 		c="-jail"
 		;;
 	Linux)
-		c="-docker -lxc -openvz -rkt -systemd-nspawn -uml -vserver"
+		c="-docker -lxc -openvz -rkt -systemd-nspawn -uml -vserver -zone"
 		;;
 	esac
 	echo $c
diff --git a/src/librc/librc.c b/src/librc/librc.c
index 9da1757f..3b6d50cd 100644
--- a/src/librc/librc.c
+++ b/src/librc/librc.c
@@ -270,6 +270,8 @@ detect_container(const char *systype _unused)
 				return RC_SYS_SYSTEMD_NSPAWN;
 		if (strcmp(systype, RC_SYS_DOCKER) == 0)
 				return RC_SYS_DOCKER;
+		if (strcmp(systype, RC_SYS_ZONE) == 0)
+				return RC_SYS_ZONE;
 	}
 	if (file_regex("/proc/cpuinfo", "UML"))
 		return RC_SYS_UML;
@@ -287,6 +289,8 @@ detect_container(const char *systype _unused)
 		return RC_SYS_RKT;
 	else if (file_regex("/proc/1/environ", "container=systemd-nspawn"))
 		return RC_SYS_SYSTEMD_NSPAWN;
+	else if (file_regex("/proc/1/environ", "container=zone"))
+		return RC_SYS_ZONE;
 	else if (exists("/.dockerenv"))
 		return RC_SYS_DOCKER;
 	/* old test, I'm not sure when this was valid. */
diff --git a/src/librc/rc.h.in b/src/librc/rc.h.in
index d2e51dc6..5276f892 100644
--- a/src/librc/rc.h.in
+++ b/src/librc/rc.h.in
@@ -318,6 +318,7 @@ bool rc_service_daemons_crashed(const char *);
 /*! @name System types
  * OpenRC can support some special sub system types, normally virtualization.
  * Some services cannot work in these systems, or we do something else. */
+#define RC_SYS_ZONE    "ZONE"
 #define RC_SYS_DOCKER  "DOCKER"
 #define RC_SYS_JAIL    "JAIL"
 #define RC_SYS_NONE    ""
-- 
2.26.2

